RFM9x.device.lib.nut

```diff
@@ -0,0 +1,437 @@
+const RFM9X_REG_RX_N_BYTES = 0x13; 
+
+const RFM9X_REG_RX_HEADER_CNT_VALUE_MSB = 0x14;
+const RFM9X_REG_RX_HEADER_CNT_VALUE_LSB = 0x15;
+const RFM9X_REG_RX_PACKET_CNT_VALUE_MSB = 0x16;
+const RFM9X_REG_RX_PACKET_CNT_VALUE_LSB = 0x17;
+
+const RFM9X_REG_MODEL_STAT = 0x18;
+const RFM9X_REG_PKT_SNR_VALUE = 0x19;
+const RFM9X_REG_PKT_RSSI_VALUE = 0x1A;
+const RFM9X_REG_RSSI_VALUE = 0x1B;
+const RFM9X_REG_HOP_CHANNEL = 0x1C;
+const RFM9X_REG_MODEM_CONFIG1 = 0x1D;
+const RFM9X_REG_MODEM_CONFIG2 = 0x1E;
+const RFM9X_REG_MODEM_CONFIG3 = 0x26;
+const RFM9X_REG_SYMB_TIMEOUT_LSB = 0x1F;
+
+const RFM9X_REG_PREAMBLE_MSB = 0x20;
+const RFM9X_REG_PREAMBLE_LSB = 0x21;
+const RFM9X_REG_PAYLOAD_LENGTH = 0x22;
+const RFM9X_REG_MAX_PAYLOAD_LENGTH = 0x23;
+const RFM9X_REG_HOP_PERIOD = 0x24;
+const RFM9X_REG_FIFO_RX_BYTE_ADDR = 0x25;
+
+
+// -------------------- Shared Registers ---------------------- //
+
+// DIO mappings are used for interrupts
+
+const RFM9X_REG_DIO_MAPPING1 = 0x40;
+const RFM9X_REG_DIO_MAPPING2 = 0x41;
+const RFM9X_REG_VERSION = 0x42;
+const RFM9X_REG_TCXO = 0x4B;
+
+// Default init constants
+const RFM9X_DEFAULT_FREQUENCY = 915000000;
+const RFM9X_DEFAULT_BANDWIDTH = "125";
+const RFM9X_DEFAULT_SF = 7;
+const RFM9X_DEFAULT_CODING_RATE = "4/5";
+const RFM9X_DEFAULT_IMPLICIT_HEADER_MODE = 0;
+const RFM9X_DEFAULT_PREAMBLE_LENGTH = 8;
+
+const RFM9X_FREQ_STEP = 61.035;
+
+// LoRa Mode
+const RFM9X_CONFIG_BYTE = 0x80; 
+
+// Interrupt flags
+enum RFM9X_FLAGS {
+    CAD_DETECTED,
+    FHSS_CHANGE_CHANNEL,
+    CAD_DONE,
+    TX_DONE,
+    VALID_HEADER,
+    PAYLOAD_CRC_ERROR,
+    RX_DONE,
+    RX_TIMEOUT
+};
+
+// Operating modes
+const RFM9X_SLEEP = 0x00;
+const RFM9X_STANDBY = 0x01;
+const RFM9X_FSTX = 0x02;
+const RFM9X_TX = 0x03;
+const RFM9X_FSRX = 0x04;
+const RFM9X_RXCONTINUOUS = 0x05;
+const RFM9X_RXSINGLE = 0x06;
+const RFM9X_CAD = 0x07;
+
+
+class RFM9x {
+    static VERSION = "0.1.0";
+
+    // -------------------- Some Tables --------------------------- //
+    
+    // This table maps bandwidth strings (in kHz) to the values which go in the 
+    // corresponding register
+    static BWTABLE = {
+        "7.8" : 0x00,
+        "10.4" : 0x01,
+        "15.6" : 0x02,
+        "20.8" : 0x03,
+        "31.25": 0x04,
+        "41.7": 0x05,
+        "62.5": 0x06,
+        "125": 0x07,
+        "250" : 0x08,
+        "500": 0x09
+    };
+    
+    // This table maps coding rate strings to the values which go in the corresponding
+    // register
+    static CRTABLE = {
+        "4/5": 0x01,
+        "4/6": 0x02,
+        "4/7": 0x03,
+        "4/8": 0x04
+    };
+
+    _spiModule = null;
+    _cs = null;
+    _irqPin = null;
+    _receiveHandler = null;
+    _isSending = false;
+    _error = false;    
+    _sendcb = null;
+    
+    constructor(spi, intPin, sendcb, cs=null) {
+        // Assume spi is already initialized, assume cs is already initialized
+        _spiModule = spi;
+        _irqPin = intPin;
+        _cs = cs;
```

Comment:
cs pin needs to be in a specific state for us to use it, and it doesn't hurt to reconfigure it, so I think most library's will configure the cs pin but require spi to be configured.